web3/providers/websocket.py

```diff
@@ -0,0 +1,67 @@
+import asyncio
+import json
+import os
+from threading import (
+    Thread,
+)
+
+from web3.exceptions import (
+    NotConfigured,
+)
+from web3.providers.base import (
+    JSONBaseProvider,
+)
+
+try:
+    import websockets
+except ImportError:
+    websockets = None
+
+
+def _start_event_loop(loop):
+    asyncio.set_event_loop(loop)
+    loop.run_forever()
+
+
+# TODO: Think of a better name
+def _get_threaded_loop():
+    new_loop = asyncio.new_event_loop()
+    t = Thread(target=_start_event_loop, args=(new_loop,))
+    # TODO: figure out if daemon is the way to go
+    t.daemon = True
```

Comment:
Ideally we want to have the life cycle of the thread and the provider be explicitly bound together such that if an instance of the provider goes away, so does the thread.  This implies to me that we need *both* the thread **and** the loop from this function.

Reply:
I had taken the same approach before submitting, However I couldn't manage to close the event_loop ad the thread successfully. I tried doing this in `__del__`  method.
I admit that I have limited knowledge about the life cycle of python objects and threads. I will give this another try. 
However, If you could give me some of guidance then It'd be very helfpful.